pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -443,6 +444,120 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
+                assets: _asIAsset(tokens),
+                maxAmountsIn: maxAmountsIn,
+                userData: abi.encode(WeightedPoolUserData.JoinKind.ADD_TOKEN, tokenAmountIn),
+                fromInternalBalance: false
+            })
+        );
+
+        _denormWeightSum = weightSumAfterAdd;
+
+        if (mintAmount > 0) {
+            _mintPoolTokens(recipient, mintAmount);
+        }
+
+        emit TokenAdded(token, normalizedWeight, tokenAmountIn);
+    }
+
+    function _validateAddToken(uint256 normalizedWeight) private view returns (uint256) {
+        _require(normalizedWeight >= WeightedMath._MIN_WEIGHT, Errors.MIN_WEIGHT);
+        // The max weight actually depends on the number of other tokens, but this is handled below
+        // by ensuring the final weights are all >= minimum
+        _require(normalizedWeight < FixedPoint.ONE, Errors.MAX_WEIGHT);
+        _require(_getTotalTokens() < _getMaxTokens(), Errors.MAX_TOKENS);
+
+        // Do not allow adding tokens if there is an ongoing or pending gradual weight change
+        _ensureNoWeightChange();
+
+        // The growth in the total weight of the pool can be easily calculated by:
+        //
+        // weightSumRatio = totalWeight / (totalWeight - newTokenWeight)
+        //
+        // As we're working with normalized weights `totalWeight` is equal to 1.
+        //
+        // We can then easily calculate the new denormalized weight sum by applying this ratio to the old sum.
+        uint256 weightSumAfterAdd = _denormWeightSum.mulUp(FixedPoint.ONE.divDown(FixedPoint.ONE - normalizedWeight));
+
+        // Now we need to make sure the other token weights don't get scaled down below the minimum
+        // normalized weight[i] = denormalized weight[i] / weightSumAfterAdd
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        for (uint256 i = 0; i < tokens.length; i++) {
+            _require(
+                _getTokenData(tokens[i]).decodeUint64(_END_DENORM_WEIGHT_OFFSET).uncompress64(_MAX_DENORM_WEIGHT).divUp(
+                    weightSumAfterAdd
+                ) >= WeightedMath._MIN_WEIGHT,
+                Errors.MIN_WEIGHT
+            );
+        }
+
+        return weightSumAfterAdd;
+    }
+
+    function _registerNewToken(IERC20 token, uint256 denormalizedWeight) private returns (IERC20[] memory) {
+        IERC20[] memory tokensToAdd = new IERC20[](1);
+        tokensToAdd[0] = token;
+
+        getVault().registerTokens(getPoolId(), tokensToAdd, new address[](1));
+
+        // Tokens array is now different
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+
+        // `_encodeTokenState` performs an external call to `token` (to get it's decimals value), however this is
+        // reentrancy safe as view functions are called in a STATICCALL context and so will revert if it modifies state.
+        _tokenState[token] = _encodeTokenState(token, denormalizedWeight, denormalizedWeight);
```

Comment:
I've broken this in [51ba378](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1131/commits/51ba3783f22f8f34f58d206ccbed573c3b18fd52) as using `_encodeTokenState` takes in a normalized weight and uses the **old** denormWeightSum when writing it to storage.

Reply:
Fixed in [41234db](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1131/commits/41234dbe6a5da09d49185de551bcb6efd54be854) as we now pass in the normalized weight along with the new `denormWeightSum` value.